worker/jobs/pull/base.py

```diff
@@ -0,0 +1,94 @@
+import ipaddress
+import re
+import requests
+from socket import gethostbyname
+from urllib.parse import urlparse
+
+
+class RemoteFileException(Exception):
+    pass
+
+
+class PullSession(requests.sessions.Session):
+    def __init__(self):
+        super().__init__()
+        self.max_redirects = 5
+        self.headers = {"User-Agent": "Stencila Hub HTTP Client"}
+
+    def get_redirect_target(self, resp):
+        # Override to run each redirect target through the malicious
+        # host check
+        url = super().get_redirect_target(resp)
+        self.check_host(url)
+        return url
+
+    def check_host(self, url):
+        if url is None:
+            return
+
+        url_obj = urlparse(url)
+
+        if url_obj.hostname is None:
+            raise TypeError('Url "{}" appears to have no hostname.'.format(url))
+
+        if is_malicious_host(url_obj.hostname):
+            raise RemoteFileException(
+                "{} is not a valid hostname.".format(url_obj.hostname)
+            )
+
+    def fetch_url(self, url, stream=False):
+        url = convert_raw_content_url(url)
+        self.check_host(url)
+        return self.get(url, stream=stream, allow_redirects=True)
+
+    def pull(self, url, sink):
+        with self.fetch_url(url, stream=True) as response:
+            with open(sink, "wb") as file:
+                for chunk in response:
+                    file.write(chunk)
+
+
+def convert_raw_content_url(url: str) -> str:
+    """Get URL to raw content from one that has website chrome around it.
+
+    For providers like Github that have a URL you would navigate to that
+    is not the raw content, convert the URL to be that of the raw
+    content.
+    """
+    github_match = re.search(
+        r"^https?://github\.com/([a-z0-9\-]+)/([a-z0-9\-_]+)/blob/([^?]+)", url, re.I
+    )
+
+    if github_match:
+        return "https://raw.githubusercontent.com/{}/{}/{}".format(
+            github_match.group(1), github_match.group(2), github_match.group(3)
+        )
+
+    hackmd_match = re.search(r"^https?://hackmd\.io/([^/?]+)", url, re.I)
+
+    if hackmd_match:
+        return "https://hackmd.io/{}/download".format(hackmd_match.group(1))
+
+    return url
+
+
+def is_malicious_host(hostname):
```

Comment:
For greater encapsulation, maybe this would be better as a `staticmethod` of `PullSession`?